### PR TITLE
Annotate a couple more marshal APIs

### LINKED ISSUE #634: Marshal.GetDelegateForFunctionPointer should be AOT unsafe
We should mark the version that takes System.Type as AOT unfriendly.

But: dataflow analysis should figure out the simple cases (the parameter is a `typeof` that we just see).

Basically do the same thing as we do for `Marshal.SizeOf`.

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
And add intrinsic handling. Fixes #634.

### CODE REVIEW

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs [hunk 1]
```diff
@@ -330,6 +330,9 @@ enum IntrinsicId
             Expression_New,
             Enum_GetValues,
             Marshal_SizeOf,
+            Marshal_PtrToStructure,
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs:333]: `StructureToPtr<T>` as well?

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs [hunk 2]
```diff
@@ -330,6 +330,9 @@ enum IntrinsicId
...
+            Marshal_DestroyStructure,
+            Marshal_GetDelegateForFunctionPointer,
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs:335]: `Marshal.GetFunctionPointerForDelegate` as well?

**Review by jkotas — APPROVED** [review]
Thank you!

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs (hunk 1 above)
Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs:333]: The generic versions can be analyzed without dataflow and we do that: https://github.com/dotnet/corert/blob/24f1edc938af9954eac9a3385958787ca64956d5/src/ILCompiler.Compiler/src/Compiler/UsageBasedInteropStubManager.cs#L88-L121

There's no nongeneric variant that takes a Type, so there's no way to make this with with the current dataflow analysis (we don't track object types).

Comment (jkotas) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs:333]: Ah ok.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.CoreRT.cs [hunk 3]
```diff
@@ -136,6 +137,7 @@ public static unsafe void DestroyStructure(IntPtr ptr, Type structuretype)
             }
         }
 
+        [RequiresDynamicCode("Marshalling code for the object might not be available")]
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.CoreRT.cs:140]: Should we include suggestion for how to fix where possible? E.g. "Use Marshal.StructurePtr<T> instead."

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs (hunk 2 above)
Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs:335]: Ah. Didn't see it in Intellisense and apparently we obsoleted and editorbrowsable(never)-hidden it.

Seems like the other non-generic version is obsolete. But people still use it. Doesn't hurt to add support for it.

Comment (jkotas) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs:335]: I think it has the same problem as `StructureToPtr`. It takes object, not the type. So we can just emit warning for it.

**Pull request merged**